src/lfu.ml

```diff
@@ -0,0 +1,120 @@
+module Make (K : sig
+  type t
+
+  val equal : t -> t -> bool
+  val hash : t -> int
+end) =
+struct
+  module H = Hashtbl.Make (K)
+
+  type key = K.t
+  type freq_cell = (int * key Dbllist.t) Dbllist.cell
+  type key_cell = key Dbllist.cell
+
+  type 'a t = {
+    value : (freq_cell * key_cell * 'a) H.t;
+    frequency : (int * key Dbllist.t) Dbllist.t;
+    cap : int;
+    stats : Stats.t;
+  }
+
+  let unsafe_v c =
+    {
+      value = H.create c;
+      frequency = Dbllist.create ();
+      cap = c;
+      stats = Stats.v ();
+    }
+
+  let v c =
+    if c <= 0 then invalid_arg "capacity must be strictly positive";
+    unsafe_v c
+
+  let stats t = t.stats
+  let is_empty t = H.length t.value = 0
+  let capacity t = t.cap
+  let size t = H.length t.value
+
+  let clear t =
+    H.clear t.value;
+    Dbllist.clear t.frequency;
+    Stats.clear t.stats
+
+  let update t k =
+    let freq_cell, key_cell, _value = H.find t.value k in
+    let freq, freq_list = freq_cell.content in
+    let freq_next, _freq_next_list = freq_cell.next.content in
+    (if freq <> freq_next - 1 then
+     let real_next_freq_list = Dbllist.create () in
+     let real_freq = freq + 1 in
+     ignore
+       (Dbllist.append_after t.frequency freq_cell
+          (real_freq, real_next_freq_list)
+         : freq_cell));
+    Dbllist.remove freq_list key_cell;
+    let _freq_next, freq_next_list = freq_cell.next.content in
+    if Dbllist.is_empty freq_list then Dbllist.remove t.frequency freq_cell;
+    let last = Dbllist.append freq_next_list k in
+    (freq_cell.next, last)
+
+  let find t k =
+    let _freq_cell, _key_cell, v = H.find t.value k in
+    let new_freq_cell, new_last_cell = update t k in
+    H.replace t.value k (new_freq_cell, new_last_cell, v);
+    v
+
+  let find_opt t k = try Some (find t k) with Not_found -> None
+
+  let mem t k =
+    try
+      ignore (find t k);
+      true
+    with Not_found -> false
+
+  let add t k v =
+    if H.length t.value = 0 then
+      let first_freq_list = Dbllist.create () in
+      let new_cell = Dbllist.append first_freq_list k in
+      let first_freq_cell = Dbllist.append t.frequency (1, first_freq_list) in
+      H.replace t.value k (first_freq_cell, new_cell, v)
+    else
```

Comment:
Have you also verified that this works if `t` is not a fresh `lfu` but a lfu that was made empty by a call to `remove`?

Reply:
Now tested in https://github.com/pascutto/cachecache/pull/8/commits/5906bd3ee5771426cf6656d650efb4bee2d5e79b :)